### PR TITLE
Fast board refactor and comments

### PR DESCRIPTION
**Pull request by barrybecker4**
@gcp, I added some comments based on your responses in #173 and did a simple refactor so that the code to print the column header labels is not duplicated. Hope you can review. I am mostly a java, javascript, and scala programmer these days, but in the distant past I used c++. The PR is mostly about me learning and trying to get familiar with using MSVC. I'd feel a bit more comfortable making changes if there were some unit tests. Maybe I will open a separate issue for that.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,13 @@ leelaz
 *.o
 *.d
 training/tf/venv
-weights.txt
\ No newline at end of file
+weights.txt
+.idea
```
Comment (gcp) [inline .gitignore]: Remove the .idea (we had a discussion that we're not going to put everyone's favorite IDE/editor in here) and duplicate the VS2015 parts to VS2017.

File: .gitignore [hunk 2]
```diff
@@ -2,4 +2,13 @@ leelaz
...
+/msvc/packages
+/msvc/VS2015/x64/Debug
+/msvc/VS2015/x64/Release
+/msvc/x64/Debug
+/msvc/x64/Release
+/msvc/leela-zero2015.VC.db
+/msvc/leela-zero2015.VC.VC.opendb
+/msvc/.vs/leela-zero2015/v14/.suo
```
Comment (gcp) [inline .gitignore]: These last 3 look unfamiliar to me.

File: src/FastBoard.h [hunk 3]
```diff
@@ -154,6 +165,7 @@ class FastBoard {
     int update_board_eye(const int color, const int i);
     int in_atari(int vertex);
     bool fast_in_atari(int vertex);
+	void print_column_labels(int size, std::string padding = " ");
```
Comment (gcp) [inline src/FastBoard.h]: nit: incorrect indent

File: src/FastBoard.cpp [hunk 4]
```diff
@@ -38,19 +38,22 @@ const int FastBoard::BIG;
 const int FastBoard::PASS;
 const int FastBoard::RESIGN;
 
+/* indicates that an empty square is never in atari. A large value that fits in a short */
+const int MAX_LIBS = 16384;  // 2^14
```
Comment (gcp) [inline src/FastBoard.cpp]: Move the definition to the header for consistency. It should be a constexpr as well. Also it's not really a max value. INF_LIBS?

File: src/FastBoard.cpp [hunk 5]
```diff
@@ -200,10 +210,13 @@ void FastBoard::reset_board(int size) {
     }
 
     m_parent[MAXSQ] = MAXSQ;
-    m_libs[MAXSQ]   = 16384;    /* we will subtract from this */
+    m_libs[MAXSQ]   = MAX_LIBS;    /* we will subtract from this */
     m_next[MAXSQ]   = MAXSQ;
 }
 
+/*
+    @return true if placing the specified colored stone at the specified position would be suicide
```
Comment (gcp) [inline src/FastBoard.cpp]: IMHO this comment is entirely superfluous.

File: src/FastBoard.cpp [hunk 6]
```diff
@@ -255,17 +268,30 @@ bool FastBoard::is_suicide(int i, int color) {
     }
 }
 
+/*
+    Liberties on a single point - called pseudo liberties.
```
Comment (gcp) [inline src/FastBoard.cpp]: If you're going to add an explanation of pseudoliberties, they should probably go with the respective datastructure in the header, rather than a random function using them. (The explanation is also not very clear, I think)

File: src/FastBoard.cpp [hunk 7]
```diff
@@ -255,17 +268,30 @@ bool FastBoard::is_suicide(int i, int color) {
...
+    Count neighbours of color c at vertex v.
+    The border of the board has fake neighours of both colors.
+*/
 int FastBoard::count_neighbours(const int c, const int v) {
     assert(c == WHITE || c == BLACK || c == EMPTY);
     return (m_neighbours[v] >> (NBR_SHIFT * c)) & 7;
 }
 
+/* 
+    @return true if playing the specified single stone (ss) is suicidal.
```
Comment (gcp) [inline src/FastBoard.cpp]: true if playing at the given location is a single-stone suicide.

File: src/FastBoard.cpp [hunk 8]
```diff
@@ -392,7 +421,10 @@ std::vector<bool> FastBoard::calc_reach_color(int col) {
     return bd;
 }
 
-// Needed for scoring passed out games not in MC playouts
+/*
+   The score is from the point of view of the black player. A negative score means white is leading.
+   @return score needed for scoring passed out games not in Monte-Carlo play-outs
```
Comment (gcp) [inline src/FastBoard.cpp]: It returns the area score. The explanation why this is needed is not related to the return value, so put it in the function description.

File: src/FastBoard.cpp [hunk 9]
```diff
@@ -427,6 +462,9 @@ int FastBoard::estimate_mc_score(float komi) {
     return bsc-wsc-((int)komi)+1;
 }
 
+/*
+    @return final Monte-Carlo play-out score
```
Comment (gcp) [inline src/FastBoard.cpp]: Again, looks like a superfluous comment.

File: src/FastBoard.cpp [hunk 10]
```diff
@@ -418,6 +450,9 @@ int FastBoard::get_stone_count() {
     return m_totalstones[BLACK] + m_totalstones[WHITE];
 }
 
+/*
+    @return estimate of the Monte-Carlo play-out score
```
Comment (gcp) [inline src/FastBoard.cpp]: That's what the function name says.

File: src/FastBoard.cpp [hunk 11]
```diff
@@ -455,15 +493,8 @@ float FastBoard::final_mc_score(float komi) {
...
-    myprintf("\n   ");
-    for (int i = 0; i < boardsize; i++) {
-        if (i < 25) {
-            myprintf("%c ", (('a' + i < 'i') ? 'a' + i : 'a' + i + 1));
-        } else {
-            myprintf("%c ", (('A' + (i-25) < 'I') ? 'A' + (i-25) : 'A' + (i-25) + 1));
-        }
-    }
-    myprintf("\n");
+	myprintf("\n");
```
Comment (gcp) [inline src/FastBoard.cpp]: nit: incorrect spacing

This problem is repeated all over. I think your editor is misconfigured. Also if you can, try avoid lines longer than 80 characters.

File: src/FastBoard.cpp [hunk 12]
```diff
@@ -565,6 +581,22 @@ void FastBoard::display_liberties(int lastmove) {
     myprintf("\n\n");
 }
 
+void FastBoard::print_column_labels(int size, std::string padding) {
+	myprintf("   ");
```
Comment (gcp) [inline src/FastBoard.cpp]: nit: inconsistent indentation with the rest of the program

padding isn't change so it can be passed by const reference.

File: src/FastBoard.cpp [hunk 13]
```diff
@@ -716,6 +748,9 @@ int FastBoard::update_board_fast(const int color, const int i, bool & capture) {
     return -1;
 }
 
+/*
+    @return true if surrounded on 4 sides by the specified color and there are enough diagonals to avoid false eye.
```
Comment (gcp) [inline src/FastBoard.cpp]: This is the good kind of comment 👍

File: src/FastBoard.cpp [hunk 14]
```diff
@@ -565,6 +581,22 @@ void FastBoard::display_liberties(int lastmove) {
     myprintf("\n\n");
 }
 
+void FastBoard::print_column_labels(int size, std::string padding) {
+	myprintf("   ");
+	for (int i = 0; i < size; i++) {
+		if (i < 25) {
```
Comment (gcp) [inline src/FastBoard.cpp]: I think Leela Zero is limited to maximum 25 sized boards so the >= 25 case can probably go.

File: src/FastBoard.cpp [hunk 15]
```diff
@@ -1000,6 +1036,14 @@ int FastBoard::string_size(int vertex) {
     return m_stones[m_parent[vertex]];
 }
 
+/*
+   Actual (real) liberties of the string the specified vertex position belongs to.
+   Implemented using union-find.
```
Comment (gcp) [inline src/FastBoard.cpp]: That's rather misleading. The calculation is done incrementally, union-find is only used for locating the parent vertex. And even so, I don't think it's the pure union find that requires recursing on find-parent.

File: src/FastBoard.cpp [hunk 16]
```diff
@@ -1000,6 +1036,14 @@ int FastBoard::string_size(int vertex) {
...
+   The difference is speed, pseudo liberties have O(1) adding and removal,
+   and real liberties have the inverse Ackermann function for that,
+   but in practice there is a significant speed difference.
```
Comment (gcp) [inline src/FastBoard.cpp]: This is just wrong. There's a difference between keeping track of the liberties and keeping track of the parent, and the latter is not the speed critical part.

File: .gitignore (hunk 1 above)
Comment (barrybecker4) [reply .gitignore]: OK. done.

File: .gitignore (hunk 2 above)
Comment (barrybecker4) [reply .gitignore]: I was going to remove them from .gitignore, but then I did not see how to exclude them from the commit using VS. I removed them and pushed from the cmd line.

File: src/FastBoard.cpp (hunk 4 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: Moved it to .h and renamed it to INF_LIBS.

File: src/FastBoard.cpp (hunk 6 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: I was basing the comment on what @fishcu said, and which you indicated was a correct description. There is no data structure for pseudoliberties that I can find - only the count_pliberties method which calls count_neighbors(EMPTY, i). I reduced the comment to just a single line until a better description can be found.

File: src/FastBoard.cpp (hunk 7 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: Updated the comment to your suggestion.

File: src/FastBoard.cpp (hunk 8 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: I removed the @return tag and updated the function description.

File: src/FastBoard.cpp (hunk 10 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: OK. I removed the comment, but it was not automatically obvious to me when I first read the code what _mc_ stood for.

File: src/FastBoard.cpp (hunk 12 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: Added the const qualifier.

File: src/FastBoard.cpp (hunk 11 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: Yes. I did not have VS configured correctly so it was mixing in tabs. I think I have it configured correctly now. All tabs replaced with spaces and I turned on visible white-space to avoid future occurrences. I will try to keep lengths below 80, but that seems pretty low, and there are many lines that exceed that in the original.

File: src/FastBoard.cpp (hunk 14 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: OK. I removed that case.

File: src/FastBoard.cpp (hunk 13 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: Glad to see I got something right.

File: src/FastBoard.cpp (hunk 15 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: I was going off of what @fishcu said. Reduced it to just a single line until a better description can be found.

File: src/FastBoard.cpp (hunk 16 above)
Comment (barrybecker4) [reply src/FastBoard.cpp]: removed it.

File: src/FastBoard.h (hunk 3 above)
Comment (barrybecker4) [reply src/FastBoard.h]: Fixed. No more tabs.

**Comment by barrybecker4** [comment]
I made changes based on review comments, and merged with the latest. It seems that a bunch of code was removed recently.

**Comment by gcp** [comment]
This unfortunately got overtaken by the large number of changes (and simplifications in FastBoard) being pulled in. Not sure how feasible it is to try another rebase.

**Comment by barrybecker4** [comment]
I tried building the next branch in Visual studio 2015 and 2017 this morning, but I am getting the following error for autogtp. Not sure how to resolve it.

1>leela-zero.vcxproj -> E:\projects\leela-zero\msvc\x64\Debug\leela-zero.exe
1>Done building project "leela-zero.vcxproj".
2>------ Build started: Project: autogtp, Configuration: Debug x64 ------
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppBuild.targets(387,5): warning MSB8028: The intermediate directory (x64\Debug\) contains files shared from another project (leela-zero.vcxproj).  This can lead to incorrect clean and rebuild behavior.
2>Performing Custom Build Tools
2>The system cannot find the file specified.
2>Performing Custom Build Tools
2>The system cannot find the file specified.
2>Moc'ing Job.h...
2>The system cannot find the path specified.
2>Moc'ing Management.h...
2>The system cannot find the path specified.
2>Moc'ing Worker.h...
2>The system cannot find the path specified.
2>Generate moc_predefs.h
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code 2.
2>Done building project "autogtp.vcxproj" -- FAILED.
========== Build: 1 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

**Comment by isty2e** [comment]
Is #583 somehow relevant to the building failure?

**Comment by gcp** [comment]
Going to close this for now. A new attempt is going to require quite some rebasing anyway.
